engine/servermaster/server.go

```diff
@@ -759,30 +754,45 @@ func (s *Server) runLeaderService(ctx context.Context) (err error) {
 		return s.gcCoordinator.Run(errgCtx)
 	})
 
-	metricTicker := time.NewTicker(defaultMetricInterval)
-	defer metricTicker.Stop()
-	leaderTicker := time.NewTicker(time.Millisecond * 200)
-	defer leaderTicker.Stop()
-	for {
-		select {
-		case <-ctx.Done():
-			// ctx is a leaderCtx actually
-			return ctx.Err()
-		case <-leaderTicker.C:
-			if !s.isEtcdLeader() {
-				log.L().Info("etcd leader changed, resigns server master leader",
-					zap.String("old-leader-name", s.name()))
-				return derrors.ErrEtcdLeaderChanged.GenWithStackByArgs()
-			}
-			err := s.jobManager.Poll(ctx)
-			if err != nil {
-				log.L().Warn("Polling JobManager failed", zap.Error(err))
-				return err
+	errg.Go(func() error {
+		defer func() {
+			s.executorManager.Stop()
+			log.L().Info("executor manager exited")
+		}()
+		s.executorManager.Start(errgCtx)
+		return nil
+	})
+
+	errg.Go(func() error {
+		metricTicker := time.NewTicker(defaultMetricInterval)
+		defer metricTicker.Stop()
+		leaderTicker := time.NewTicker(time.Millisecond * 200)
+		defer leaderTicker.Stop()
+		for {
+			select {
+			case <-errgCtx.Done():
+				// errgCtx is a leaderCtx actually
+				return perrors.Trace(errgCtx.Err())
+			case <-leaderTicker.C:
+				if !s.isEtcdLeader() {
```

Comment:
Would it be better if `runLeaderService` is canceled from outside if the Etcd leadership is lost?

Reply:
Yes, move the cancel logic to `leaderLoop`, it is a logic related to server manager, instead of session or campaign library.
Besides we don't watch the etcd leader change event, so only check it periodically.